lib/FrontendTool/FrontendTool.cpp

```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
 }
 }
 
+namespace {
+
+// The most interesting/tricky thing here is _which_ paths get recorded in
+// the trace file as dependencies. It depends on how the module was synthesized.
+// The key points are:
+//
+// 1. Paths to swiftmodules in the module cache or in the prebuilt cache are not
+//    recorded - Precondition: the corresponding path to the swiftinterface must
+//    already be present as a key in pathToModuleDecl.
+// 2. swiftmodules next to a swiftinterface are saved if they are up-to-date.
+void computeSwiftModuleTraceInfo(
+    const SmallPtrSetImpl<ModuleDecl *> &importedModules,
+    const llvm::DenseMap<StringRef, ModuleDecl *> &pathToModuleDecl,
+    const DependencyTracker &depTracker,
+    StringRef prebuiltCachePath,
+    std::vector<SwiftModuleTraceInfo> &traceInfo) {
+
+  SmallString<256> buffer;
+
+  std::string errMsg;
+  llvm::raw_string_ostream err(errMsg);
+
+  auto errorUnexpectedPath = [&err, &pathToModuleDecl]() {
+    err << "The module <-> path mapping we have is:\n";
+    for (auto &m: pathToModuleDecl)
+      err << m.second->getName() << " <-> " << m.first << '\n';
+    llvm::report_fatal_error(err.str());
+  };
```

Comment:
I really don't like this "print more stuff to the error string and then error" idiom, doubly so when it's relying on capturing the error string stream. Can you make a PrettyStackTrace subclass for dumping this mapping instead? (This can happen in a follow-up PR though.)

Reply:
I will make this change in a cleanup PR later.